dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/common/info/OpenTelemetryInfoFactoryImpl.java

```diff
@@ -0,0 +1,331 @@
+        ResourceBuilder builder = resource.toBuilder();
+        builder.put(ResourceAttributes.SERVICE_NAME, getServiceName(c));
+        return builder.build();
+    }
+
+    //Uses application name if the user has not given configured service.name resource attribute
+    private static String getServiceName(ConfigProperties c) {
+        String appName = c.getString(SERVICE_NAME_PROPERTY);
+        ComponentMetaData cmd = ComponentMetaDataAccessorImpl.getComponentMetaDataAccessor().getComponentMetaData();
+        if (appName == null) {
+            if (cmd != null) {
+                appName = cmd.getModuleMetaData().getApplicationMetaData().getName();
+            }
+        }
+        return appName;
+    }
+
+    //Interfaces and private classes only relevent to this factory.
+
+    public interface OpenTelemetrySdkBuilderSupplier {
+        public AutoConfiguredOpenTelemetrySdkBuilder getPartiallyConfiguredOpenTelemetrySDKBuilder();
+    }
+
+    /*
+     * There are two race conditions we need to protect against.
+     *
+     * On app startup/shutdown we need to swap the holder associated with the application.
+     * We protect against race conditions here by using an AtomicReference.
+     *
+     * Within the context of an application's lifecycle we need to ensure OpenTelemetryInfo
+     * is only created once. OpenTelemetryInfoReference handles this.
+     */
+    private class OpenTelemetryInfoReference extends AtomicReference<OpenTelemetryInfoWrappedSupplier> {
+
+        private static final long serialVersionUID = -4884222080590544495L;
+    }
+
+    private class OpenTelemetryInfoWrappedSupplier {
+        private final Consumer<OpenTelemetryInfo> consumer;
+        private Supplier<OpenTelemetryInfo> supplier;
+        private volatile OpenTelemetryInfo openTelemetryInfo = null;
+
+        public OpenTelemetryInfoWrappedSupplier(Consumer<OpenTelemetryInfo> consumer, Supplier<OpenTelemetryInfo> supplier) {
+            this.consumer = consumer;
+            this.supplier = supplier;
+        }
+
+        public OpenTelemetryInfo get() throws InterruptedException, ExecutionException {
+            if (openTelemetryInfo != null) {
+                return openTelemetryInfo;
+            }
+
+            synchronized (this) {
+                if (openTelemetryInfo != null) {
+                    return openTelemetryInfo;
+                }
+
+                openTelemetryInfo = supplier.get();
+                return openTelemetryInfo;
+            }
+        }
+
+        public boolean closeAndDisposeIfCreated() {
+            synchronized (this) {
+                supplier = OpenTelemetryInfoFactoryImpl::createDisposedOpenTelemetryInfo;
+                if (openTelemetryInfo != null) {
+                    consumer.accept(openTelemetryInfo);
+                    return true;
+                }
+                return false;
+            }
+        }
```

Comment:
I don't think this needs to return a boolean any more. If it does, the meaning of the return value should be documented.